### PR TITLE
Linting and Code Scanning

### PR DESCRIPTION
**Pull request by faddat**
## note

This PR addresses a number of code quality issues.

Per @fragwuerdig request, it no longer contains other PRs and is a freestanding PR into the main branch that lints it. 

## changes 

- add code quality tooling that was removed by l1 team
- begin cleanup (again).  Linting makes issues more obvious, gives them fewer places to hide
- Delete CODEOWNERS
- fix
- continue cleaning
- 144 quality issues that were resolved motnhs ago to re-resolve before can confidently debug
- 135 code quality issues that were resolved motnhs ago to re-resolve before can confidently debug
- 107 lint issssues I fixed months ago remaining
- 93 lint issues fixed months ago remaining
- 89 lint issues fixed months ago remaining
- 82 lint issues fixed months ago remaining
- 58 lint issues I fixed months ago remaining
- 28 lint issues that I fixed months ago, which were readded to the repo for no reason, remanining
- there.  Now we are back where we were 3 months ago.
- 100%
- go install
- linted
- fix test_utils.go
- remove exclusions
- don't use deprecated

## Static Analysis

- codeql added
- codacy added

## Summary of changes

This PR lints the codebase so that issues can be discovered more easily.  For 
example, the testnet code in the cmd folder was using a deprecated call. 

It also introduces a set of tooling that increases the rigor of automated testing.

## Report of required housekeeping

- [ ] Github issue OR spec proposal link
- [ ] Wrote tests
- [ ] Updated API documentation (client/lcd/swagger-ui/swagger.yaml)
- [ ] Added a relevant changelog entry: clog add [section] [stanza] [message]

----

## (FOR ADMIN) Before merging

- [ ] Added appropriate labels to PR
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" (coding standards)
- [ ] Confirm added tests are consistent with the intended behavior of changes
- [ ] Ensure all tests pass

### CODE REVIEW

**Comment by faddat** [comment]
Since this has everything in one place, merge it into #176, then do a check on #176.

It isn't worth spending the time to go through 

* #176 

... until it passes basic automated checks, and does not use deprecated code.

**Comment by fragwuerdig** [comment]
> This PR now contains @fragwuerdig's
> 
> * [V3 Upgrade Handling #177](https://github.com/classic-terra/core/pull/177)
> 
> I don't recommend developing the upgrade code separately from the changes, and I do not recommend using unlinted code that hasn't had extensive static analysis done on it. THis just makes the people involved do more work, with less result.

Hm. That wasn't finished yet. Should have marked it as work in progress I guess...

**Comment by faddat** [comment]
@fragwuerdig the idea is that it should be a part of v0.45.13-proposal, so that the stuff in v0.45.13-proposal is always tested with your changes in place.

**Comment by faddat** [comment]
* https://github.com/cosmos/cosmos-sdk/blob/release/v0.45.x/go.mod

**Comment by faddat** [comment]
To All: why would we revert to a deprecated version of go as suggested here:

* #160 

That doesn't seem wise at all.

**Comment by faddat** [comment]
@fragwuerdig so, basically, from here, you can safely resume work on the changes that you were working on.  They'll be checked far more strictly.

**Comment by ZaradarBH** [comment]
> You should not use go 1.18, it has reached end of life, see the SDK 46 release notes:
> 
> https://github.com/cosmos/cosmos-sdk/releases/tag/v0.46.11
> 
> You should not use go 1.18
> 
> You should use go 1.19 or go 1.20
> 
> I recommend go 1.20.

I have already told you that we will not be upgrading to 1.18 in this release. Also you still have not told me if you have tested that you can actually sync the Columbus-5 genesis in a network consisting entirely of validators using Go 1.20. For all we know it still might fail and in that cause bumping to 1.20 is a breaking change and thus a horrible idea, until we can identity the source of the none-determinism and fix it. So regardless of the LTS support on 1.18, we stick with 1.18 for now and that is the TLDR on the matter. Accept it or your PR will not get merged. End of story

**Comment by faddat** [comment]
Hi Tobias, I believe that you asked for my review, and now you have it.

**Comment by ZaradarBH** [comment]
> @ZaradarBH Is there a script available to run the chain state through a validator so that this can be tested?

No. We are working on finalizing the https://github.com/classic-terra/terra-operator beta, @nghuyenthevinh2000 is currently working on the relayer CRD we need for IBC, but once that is done we should be able to do a lot more complex integration testing that can be largely automated as we needed 7 days to test  the Go 1.19+ breakage issue after @faddat reported the genesis sync issue to us the first time as he was as fare as I know the first person to encounter this issue. Which makes sense since we didnt bump to Go 1.2 until the October PR last year which means only a few people have tested syncing with updated Go dependencies due to the time consuming nature of these types of tests

**Comment by ZaradarBH** [comment]
> Hi Tobias, I believe that you asked for my review, and now you have it.

Great. Thank you. I have commented on most of the things I can reason about. The more reviews we get, the better the code becomes and as I hope you can see from my comments security is a big deal for us, however I don't believe the argument we had over token centralization on Twitter can really ever be solved technically without something like https://docs.cosmos.network/main/architecture/adr-014-proportional-slashing which seems stuck in R&D for over 2 years.

Thus it make a lot more sense for us to focus on the things we can fix. Like all the CVEs in my postmortum doc and the ongoing process of patching supply-chain-attack vectors and potentially getting out of NodeJS and NPM all together

**Comment by faddat** [comment]
So hi there Tobias, you asked me an impossible question that you don't even have the answer to, and rejected my pull request, which is based on best practices, based on the impossible question you asked me.

My review, based on the simple best practice of "we should not use a deprecated language runtime" remains my review.

This is also the recommendation of the cosmos SDK team. They recommend using either 1.19 or 1.20, but never both. That's how you avoid the non-determinism issues, you use exactly the same language runtime across the entire validator set.

The other issues that you are describing, concerning syncing from Genesis, exist because there have been undocumented changes to the software that do not necessarily have a clear activation height.  These aren't related issues. 

Proportional slashing will never solve the incorrect belief that it is safe for a single operator to run 33% of vote power, and proportional slashing doesn't make sibyl/VaaS safe.

**Comment by faddat** [comment]
@fragwuerdig hey man, I just saw your tweet, you'd like me to take the changes from 

* #177 
* #176 

Out?

**Comment by ZaradarBH** [comment]
> So hi there Tobias, you asked me an impossible question that you don't even have the answer to, and rejected my pull request, which is based on best practices, based on the impossible question you asked me.
> 
> My review, based on the simple best practice of "we should not use a deprecated language runtime" remains my review.
> 
> This is also the recommendation of the cosmos SDK team. They recommend using either 1.19 or 1.20, but never both. That's how you avoid the non-determinism issues, you use exactly the same language runtime across the entire validator set.
> 
> The other issues that you are describing, concerning syncing from Genesis, exist because there have been undocumented changes to the software that do not necessarily have a clear activation height. These aren't related issues.
> 
> Proportional slashing will never solve the incorrect belief that it is safe for a single operator to run 33% of vote power, and proportional slashing doesn't make sibyl/VaaS safe.
> 
> * Every single chain team has had the same finding. Don't mix go run times, and then everything is fine.
> * Since go 1.18 is deprecated that leaves two viable choices:
>   
>   * 1.19
>   * 1.20
> 
> If you'd like to do engineering work based on best practices and standards, you need to choose between one of those two versions of the go language runtime, and you need to make sure that they don't mix.
> 
> I can't help the best practices and clear findings of numerous chain teams, they simply are what they are.
> 
> * [1 validator set should use 1 major release of golang cosmos/cosmos-sdk#13976](https://github.com/cosmos/cosmos-sdk/issues/13976)
> * [Bump go to 1.19 or 1.20 cosmos/gaia#2282](https://github.com/cosmos/gaia/issues/2282)
> 
> Now if you wanted to dig deeper into this, what you would find is that there are changes to the gzip library in go. And that this can cause issues with state sync.
> 
> Anecdotally there are issues with consensus, but those anecdotal consensus issues only occur when two or more run times are mixed, and it is anecdotal, we are not certain yet.

Ok let me try to be brief and as civil as I can. Every single enterprise on the planet has a system somewhere that uses a deprecated technology. Its a simple matter of fact that you cannot and should not just upgrade dependencies because you can. You do it when A) there is a business need or B) when there is a security concern. Now there is no concrete business need to upgrade to Go 1.18 and there is as such no security concern that does not already exist in the network that gives cause for us to rank this very high compared to other concerns. 

Now to conclude this for the final time, we are not doing to Go upgrade in this release. That does not mean we are never going to do it. But we have to focus on what we are currently doing, to many changes at once benefits no one. So while I appreciate the fact that you like to be principled about things, because you seem to mentally side-step the efforts that goes into doing what your proposing, I do not have the same luxury and part of my responsible is to ensure the priorities of our network first and that of the cosmos second. Now if you want to help, be constructive, listen to what people tell you and accept that in this reality VHS won, not Betamax, not because VHS was better tech, but it was more aligned with what the people desired.

**Comment by faddat** [comment]
Tobias, seems Til is going to merge this. 

Tobias there is no "cosmos agenda" here.  This PR just improves lunc and no other chain. 

Tobais, please don't promote the use of deprecated code.  Thank you.

**Comment by ZaradarBH** [comment]
> Tobias, seems Til is going to merge this.
> 
> Tobias there is no "cosmos agenda" here. This PR just improves lunc and no other chain.
> 
> Tobais, please don't promote the use of deprecated code. Thank you.

I am not promoting the use of deprecated code. As a team we have agreed to do things in a certain order and since you decided to not work as a team, that means you will have simply respect the consensus decision like I am.

**Comment by faddat** [comment]
Tobias, @fragwuerdig asked me to make changes to this PR, and target the main branch, so I did.

**Comment by ZaradarBH** [comment]
> Tobias, @fragwuerdig asked me to make changes to this PR, and target the main branch, so I did.
> 
> I am not on your team and I do not work for you. I work for Notional's delegators, and myself, in that order, in this context, and I've chosen to fulfil a reqest from @fragwuerdig. That's now done.
> 
> https://twitter.com/gadikian/status/1636138315609960451

I know, you decide to leave it because you could not work as a team. I respect that, your better off on your own, but Till does work as a team and I guarantee you that he will not approve a PR with a Go 1.19+ reference in it until we have all gained consensus on bumping Go, because we already spent enough time as a team debugging the issue and we have decided, again as a team, to leave version 1.18 in place for now. Thats the deal, take it or leave it. Either way we will be happy :pray:

**Comment by faddat** [comment]
Standards, like those surrounding deprecation notices, make decisions like these needless.

**Comment by ZaradarBH** [comment]
> Standards, like those surrounding deprecation notices, make decisions like these needless.

Go to sleep Jacob

**Comment by faddat** [comment]
I don't think that's really in the purview of the software.

File: .github/CODEOWNERS
Comment (fragwuerdig) [inline .github/CODEOWNERS:1]: like: ???

can we remove that again?

File: .github/dependabot.yml
Comment (fragwuerdig) [inline .github/dependabot.yml:1]: I don't think we need any bot watching for CometBFT deps. CometBFT is not (yet) part of our Classic codebase...

File: .github/workflows/proto-buf-publisher.yml
Comment (fragwuerdig) [inline .github/workflows/proto-buf-publisher.yml:1]: why is that needed?

File: x/msgauth/legacy/v04/types.go
Comment (fragwuerdig) [inline x/msgauth/legacy/v04/types.go:1]: I don't understand why this comment is removed. It increases readability, doesn't it?

File: .github/CODEOWNERS
Comment (fragwuerdig) [reply .github/CODEOWNERS:1]: ok. get it.

File: custom/auth/ante/ante_test.go
Comment (fragwuerdig) [inline custom/auth/ante/ante_test.go:1]: no - this actually decreases readability. use `authtypes` instead of `types`.

File: x/oracle/types/genesis.go
Comment (fragwuerdig) [inline x/oracle/types/genesis.go:1]: I'd prefer camelCase, because it's consistent with the other arguments in that function. This change decreases readability.

File: .github/workflows/proto-buf-publisher.yml
Comment (faddat) [reply .github/workflows/proto-buf-publisher.yml:1]: There are custom .proto files in lunc

File: go.mod
[the remark this answers is not in the document]
Comment (faddat) [reply go.mod:1]: I made one, @ZaradarBH closed it. 

#179

File: custom/auth/ante/ante_test.go
Comment (faddat) [reply custom/auth/ante/ante_test.go:1]: agree.

File: x/oracle/types/genesis.go [hunk 1]
```diff
@@ -12,7 +12,7 @@ func NewGenesisState(
 	feederDelegations []FeederDelegation, missCounters []MissCounter,
 	aggregateExchangeRatePrevotes []AggregateExchangeRatePrevote,
 	aggregateExchangeRateVotes []AggregateExchangeRateVote,
-	TobinTaxes []TobinTax,
+	tobinTaxes []TobinTax,
```
Comment (faddat) [inline x/oracle/types/genesis.go:15]: this is the difference between a public and private variable. tobinTaxes is in fact camelCase

File: custom/auth/ante/ante_test.go
Comment (faddat) [reply custom/auth/ante/ante_test.go:1]: addressed.

File: x/msgauth/legacy/v04/types.go
Comment (faddat) [reply x/msgauth/legacy/v04/types.go:1]: it does.

Comment (faddat) [reply x/msgauth/legacy/v04/types.go:1]: that's been removed from the file though.

**Comment by nghuyenthevinh2000** [comment]
@faddat 
it is best that you make another pr for github workflow. This one is reserved for linting.

**Comment by faddat** [comment]
Per L1TF request, I've used the go.mod and go.sum from the main branch.

I think that go.mod should be restructured. I made this PR:

* #179 

It was closed.

**Comment by faddat** [comment]
@nghuyenthevinh2000 this is the main branch bud


Screenshot 2023-03-19 at 2 19 37 AM

**Comment by nghuyenthevinh2000** [comment]
> @nghuyenthevinh2000 this is a screenshot of the main branch bud
> 
> the current workflow **does not work**
> 
> commit is:
> 
> [ff2e40b](https://github.com/classic-terra/core/commit/ff2e40b0911a88f6c1b87d1537af9c2fc2ee8229)
> 
> Screenshot 2023-03-19 at 2 19 37 AM

there is only one failing and it is docker hub authorization. Also on: https://github.com/classic-terra/core/commit/ff2e40b0911a88f6c1b87d1537af9c2fc2ee8229

Screenshot 2023-03-19 at 02 24 44

The reason for liveness test existence:

For a distributed system, the FLP theorem dictates properties of a consensus network
* Agreement: All correct nodes decide for the same value.
* Termination: All non-faulty processes eventually decide on a value in finite time.
* Validity: The decision value must be the input value of a node.

liveness test is an __overall attack__ on FLP theorem by randomly shut down nodes.

```
Lemma 16.13. If a configuration tree contains a critical configuration, crashing
a single node can create a bivalent leaf; i.e., a crash prevents the algorithm from
reaching agreement.
```

More information here: https://disco.ethz.ch/courses/hs21/distsys/lnotes/chapter16.pdf

**Comment by nghuyenthevinh2000** [comment]
https://github.com/rokroskar/workflow-run-cleanup-action seems to be deprecated, precise change to this is needed

**Comment by faddat** [comment]
The maintainers are allowed to push to this branch, you know.  

I do think it is a very significant improvement in many areas.  

Please feel welcome to fix the liveness test, I've brought it back.

**Comment by faddat** [comment]
Should really not have big red X on main.  I'm going to kinda just step back.

ikyk I know what the liveness test is.  

Do refactor go.mod, I've never seen anything like its current state anywhere before.

**Comment by faddat** [comment]
@nghuyenthevinh2000 you mean, it's been deprecated for months and months and months, and no one checked?

yeah welcome to my world.  

good job checking

**Comment by nghuyenthevinh2000** [comment]
Tobias choice for golang 1.18 over newer go version seems due to practice in banking: "If it ain’t broke, don’t fix it"

https://www.tpr.org/technology-entrepreneurship/2019-05-23/how-cobol-still-powers-the-global-economy-at-60-years-old

The COBOL language has been powering the banking system for too long time but it is stable.

If we apply this context to Terra - Classic, a system meant to be durable like banking, current go 1.18 is functioning and best not to touch it. Moving version upward introduces potential break.

Of course, not moving go version upward is not a good choice also. It is best to leave the choice of go version in upgrade tesnet phase. The current phase is development.

**Comment by faddat** [comment]
done with this for now, you have access to push to this branch, so, feel free to smooth whatever you'd like.

**Comment by nghuyenthevinh2000** [comment]
> done with this for now, you have access to push to this branch, so, feel free to smooth whatever you'd like.

thanks, I will have a look

**Comment by faddat** [comment]
Final thought -- l1tf -- supposed to maintain the chain.

The idea that this is out of scope, seems rather.... bad.

Either redefine the scope or announce you're not really maintaining the chain (obvs I recommend that you actually do maintainership work)

**Comment by faddat** [comment]
@nghuyenthevinh2000 sir, when google deprecates a language runtime really, really early, they have reasons for doing it. 

Check the CVE's 🖖 

(no seriously man check them)

**Comment by faddat** [comment]
* Code scanning results / Spectral (reported by Codacy) Successful in 26s — 1,215 new alerts
* Code scanning results / Jshint (reported by Codacy) Successful in 27s — 5,000 new alerts
* Code scanning results / CodeQL Successful in 5s — 53 new alerts, 4 fixes
* Code scanning results / Pylint (reported by Codacy) Successful in 24s — 89 new alerts

**Comment by faddat** [comment]
https://openssf.org/

File: .github/dependabot.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
+version: 2
```
Comment (ZaradarBH) [inline .github/dependabot.yml:1]: See my other comment about dependabot

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -0,0 +1,37 @@
+---
```
Comment (ZaradarBH) [inline .github/workflows/build.yml:1]: We are in the process of building a new CI/CD pipeline. So most of the assets will most likely end up being deprecated

**Review by ZaradarBH — COMMENTED** [review]
You will need to merge conflicts from the other PRs and I think there is some overlap to your other PRs.

**Comment by fragwuerdig** [comment]
@ZaradarBH https://endoflife.date/go

IMG_20230320_172955

**Comment by fragwuerdig** [comment]
> > @ZaradarBH https://endoflife.date/go
> > IMG_20230320_172955
> > However, I found no official sources on go 1.18 deprecation... (Like, from Google)
> 
> I know about this site. But just to put it into context, just because I state a perceived "ideal" as issued by us as developers. Do you really think businesses and investors care about that? In that light, do you think Google prioritizes ideals over money?
> 
> Kubectl follows the same general engineering principle with only having two-version compatibility between kubectl and the version of k8s your running in a target cluster. This does not mean that older versions are automatically deprecated and AWS have to update all their managed EKS instances every 6 months. Simply that they are no longer accepting RFCs.
> 
> So find me an official source from Google that states that everyone needs to migrate away from 1.18 under fear of death, then we can talk about re-prioritizing it.

I fully support you with this. Chain stability goes first. go 1.19 has proven to be instable with this chain. I just want to throw it into this context without offending you. Because it's important to bear in mind and we should move to the next go version. Although it won't happen with this PR.

**Comment by faddat** [comment]
@ZaradarBH sir, it really doesn't seem like I'm offended -- and I'm not.

Then again, you need to live with this:

* https://classic-agora.terra.money/t/falsehoods-promoted-by-the-l1-task-force/49913/30

"Jacob why'd you make these PR's then?"

My company's brand is associated here and I am not in support of how Luna Classic is being developed.

The chain was unstable not because of go 1.19, but because it mixed go 1.18 and 1.19 and likely 1.20 as well.

File: x/treasury/keeper/test_utils.go [hunk 4]
```diff
@@ -296,7 +298,7 @@
 	return msg
 }
 
-func setupValidators(t *testing.T) (TestInput, sdk.Handler) {
+func setupValidators(t *testing.T) TestInput {
```
[the remark this answers is not in the document]
Comment (faddat) [reply x/treasury/keeper/test_utils.go:301]: false postive

**Comment by ZaradarBH** [comment]
> @ZaradarBH sir, it really doesn't seem like I'm offended -- and I'm not.
> 
> Then again, you need to live with this:
> 
> * https://classic-agora.terra.money/t/falsehoods-promoted-by-the-l1-task-force/49913/30
> 
> "Jacob why'd you make these PR's then?"
> 
> The culture we're building at Notional doesn't permit validating a chain and letting it slide.
> 
> The chain was unstable not because of go 1.19, but because it mixed go 1.18 and 1.19 and likely 1.20 as well.

Jacob your list of supposed "falsehood" is pointless and largely self-denial on your part. If you want to extend this into the legal realm I am more then  willing and able to take this all the way. However I would advice against it because quiet frankly the internet is rife with examples of your "instability", so focus on what matters or end up getting ignored.

**Comment by nghuyenthevinh2000** [comment]
closing in favor of #270
